content/doc/developer/plugin-development/injecting-environment-variables.adoc

```diff
@@ -0,0 +1,67 @@
+---
+layout: developerguide
+title: How to inject environment variables
+---
+
+Injecting some env variables allows to communicate steps with shell scripts and another
+steps. To achieve it, just follow this guide.
+We will add `EXTERNAL_SYSTEM_ID` environment variable in Pipeline compatible way.
+
+### Using implementation of SimpleBuildStep
+
+To be compatible both with old-fashion Freestyle jobs and brand new Pipeline, your step should
+implement `SimpleBuildStep` interface, which provides `perform(Run<?,?>, FilePath, Launcher, TaskListener)` method.
+
+First of all we should create specific implementation of `hudson.model.Action`. This will allow
+us to attach some information to the `Run` instance.
+
+Our implementation should store all required data for populating environment variable.
+
+[source,java]
+----
+@Value <1>
+public class ExternalSystemLaunchedAction implements Action {
+    String name; <2>
+    String externalId; <3>
+}
+----
+<1> We use https://projectlombok.org/[Project lombok] to simplify examples
```

Comment:
I am not convinced this is a good idea. A bunch of trivial getters/setters seem easier to understand to those lacking knowledge of Lombok.

Reply:
don't think that bunch of boilerplate code is better in any place

Reply:
I would strongly advise against use of Lombok in any Jenkins code. Plain old Java is much easier to understand and debug.

In this case, `public` fields are fine. You are not defining an API anyway.

Reply:
Hmm, how often you need to debug pojo getters and setters? Public fields are EVIL, don't do that ever! Or move to something like golang or python.